testing/rest.go

```diff
@@ -0,0 +1,231 @@
+				key, e, a))
+			continue
+		}
+	}
+
+	return errs
+}
+
+// AssertHasHeader compares the header values and identifies if the actual
+// header set includes all values specified in the expect set. Emits a testing
+// error, and returns false if the headers are not equal.
+func AssertHasHeader(t T, expect, actual http.Header) bool {
+	t.Helper()
+
+	if err := HasHeader(expect, actual); err != nil {
+		for _, e := range err.(errors) {
+			t.Error(e)
+		}
+		return false
+	}
+	return true
+}
+
+// HasHeaderKeys validates that header set contains all keys expected. Returns
+// an error if a header key is not in the header set.
+func HasHeaderKeys(keys []string, actual http.Header) error {
+	var errs errors
+	for _, key := range keys {
+		if vs := actual.Values(key); len(vs) == 0 {
+			errs = append(errs, fmt.Errorf("expect %v key in %v", key, actual))
+			continue
+		}
+	}
+	return errs
+}
+
+// AssertHasHeaderKeys validates that header set contains all keys expected.
+// Emits a testing error, and returns false if the headers are not equal.
+func AssertHasHeaderKeys(t T, keys []string, actual http.Header) bool {
+	t.Helper()
+
+	if err := HasHeaderKeys(keys, actual); err != nil {
+		for _, e := range err.(errors) {
+			t.Error(e)
+		}
+		return false
+	}
+	return true
+}
+
+// NotHaveHeaderKeys validates that header set does not contains any of the
+// keys. Returns an error if a header key is found in the header set.
+func NotHaveHeaderKeys(keys []string, actual http.Header) error {
+	var errs errors
+	for _, k := range keys {
+		if vs := actual.Values(k); len(vs) != 0 {
+			errs = append(errs, fmt.Errorf("expect %v key not in %v", k, actual))
+			continue
+		}
+	}
+	return errs
+}
+
+// AssertNotHaveHeaderKeys validates that header set does not contains any of
+// the keys. Emits a testing error, and returns false if the header contains
+// the any of the keys equal.
+func AssertNotHaveHeaderKeys(t T, keys []string, actual http.Header) bool {
+	t.Helper()
+
+	if err := NotHaveHeaderKeys(keys, actual); err != nil {
+		for _, e := range err.(errors) {
+			t.Error(e)
+		}
+		return false
+	}
+	return true
+}
+
+// QueryItem provides an escaped key and value struct for values from a raw
+// query string.
+type QueryItem struct {
+	Key   string
+	Value string
+}
+
+// ParseRawQuery returns a slice of QueryItems extracted from the raw query
+// string. The parsed QueryItems preserve escaping of key and values.
+//
+// All + escape characters are replaced with %20 for consistent escaping
+// pattern.
+func ParseRawQuery(rawQuery string) (items []QueryItem) {
+	for _, item := range strings.Split(rawQuery, `&`) {
+		parts := strings.SplitN(item, `=`, 2)
+		var value string
+		if len(parts) > 1 {
+			value = parts[1]
+		}
+		items = append(items, QueryItem{
+			// Go Query encoder escapes space as `+` whereas smithy protocol
+			// tests expect `%20`.
+			Key:   strings.ReplaceAll(parts[0], `+`, `%20`),
+			Value: strings.ReplaceAll(value, `+`, `%20`),
```

Comment:
Do we need to address this in our encoder?

Reply:
Most likely i think so. I'm assuming it is using the default `Encode` method on uri.Value, which does `+` space escaping. Probably need to make a second pass after encode and replace those values with `%20`. I think the sigv4 spec has some expectations with this as well.